<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Record-Linkage" data-toc-modified-id="Record-Linkage-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Record Linkage</a></span><ul class="toc-item"><li><span><a href="#The-Principles-of-Record-Linkage" data-toc-modified-id="The-Principles-of-Record-Linkage-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>The Principles of Record Linkage</a></span></li><li><span><a href="#Linking-Patents-to-Grants:-individual-and-organization-levels" data-toc-modified-id="Linking-Patents-to-Grants:-individual-and-organization-levels-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Linking Patents to Grants: individual and organization levels</a></span></li><li><span><a href="#Python-Setup" data-toc-modified-id="Python-Setup-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Python Setup</a></span></li></ul></li><li><span><a href="#Data-Exploration:-Know-Your-Data" data-toc-modified-id="Data-Exploration:-Know-Your-Data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data Exploration: Know Your Data</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Individual-level-links:-inventor-(patent)-<->-principal-investigator-(grant)" data-toc-modified-id="Individual-level-links:-inventor-(patent)-<->-principal-investigator-(grant)-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Individual-level links: inventor (patent) &lt;-&gt; principal investigator (grant)</a></span><ul class="toc-item"><li><span><a href="#PatentsView:-inventor-names-on-patents" data-toc-modified-id="PatentsView:-inventor-names-on-patents-2.0.1.1"><span class="toc-item-num">2.0.1.1&nbsp;&nbsp;</span>PatentsView: inventor names on patents</a></span></li><li><span><a href="#FedRePORTER:-principal-investigator-names-on-grants" data-toc-modified-id="FedRePORTER:-principal-investigator-names-on-grants-2.0.1.2"><span class="toc-item-num">2.0.1.2&nbsp;&nbsp;</span>FedRePORTER: principal investigator names on grants</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Data-Cleaning" data-toc-modified-id="Data-Cleaning-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Cleaning</a></span><ul class="toc-item"><li><span><a href="#The-Importance-of-Pre-Processing" data-toc-modified-id="The-Importance-of-Pre-Processing-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>The Importance of Pre-Processing</a></span><ul class="toc-item"><li><span><a href="#Individual-level-links" data-toc-modified-id="Individual-level-links-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Individual-level links</a></span><ul class="toc-item"><li><span><a href="#Clean-Inventor-Data" data-toc-modified-id="Clean-Inventor-Data-3.1.1.1"><span class="toc-item-num">3.1.1.1&nbsp;&nbsp;</span>Clean Inventor Data</a></span></li><li><span><a href="#Clean-Grant-Data" data-toc-modified-id="Clean-Grant-Data-3.1.1.2"><span class="toc-item-num">3.1.1.2&nbsp;&nbsp;</span>Clean Grant Data</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Deterministic-Matching" data-toc-modified-id="Deterministic-Matching-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Deterministic Matching</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Individual-level-links" data-toc-modified-id="Individual-level-links-4.0.1"><span class="toc-item-num">4.0.1&nbsp;&nbsp;</span>Individual-level links</a></span><ul class="toc-item"><li><span><a href="#REGEX-CHEATSHEET" data-toc-modified-id="REGEX-CHEATSHEET-4.0.1.1"><span class="toc-item-num">4.0.1.1&nbsp;&nbsp;</span>REGEX CHEATSHEET</a></span></li><li><span><a href="#EXAMPLES" data-toc-modified-id="EXAMPLES-4.0.1.2"><span class="toc-item-num">4.0.1.2&nbsp;&nbsp;</span>EXAMPLES</a></span></li></ul></li></ul></li><li><span><a href="#References-and-Further-Readings" data-toc-modified-id="References-and-Further-Readings-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>References and Further Readings</a></span><ul class="toc-item"><li><span><a href="#Record-Linkage" data-toc-modified-id="Record-Linkage-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Record Linkage</a></span></li><li><span><a href="#Parsing" data-toc-modified-id="Parsing-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Parsing</a></span></li><li><span><a href="#Regular-Expression" data-toc-modified-id="Regular-Expression-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>Regular Expression</a></span></li><li><span><a href="#Record-Linkage-Python-package" data-toc-modified-id="Record-Linkage-Python-package-4.1.4"><span class="toc-item-num">4.1.4&nbsp;&nbsp;</span>Record Linkage Python package</a></span></li></ul></li></ul></li></ul></div>

Contributors:
Rayid Ghani, Frauke Kreuter, Julia Lane, Brian Kim, Adrianne Bradford, Alex Engler, Nicolas Guetta Jeanrenaud, Graham Henke, Daniela Hochfellner, Clayton Hunter, Avishek Kumar, Jonathan Morgan, Ekaterina Levitskaya.

# Record Linkage

This notebook will provide you with an instruction into Record Linkage using Python. Upon completion of this notebook you will be able to apply record linkage techniques using the `recordlinkage` package to combine data from different sources in Python. 
It will lead you through all the steps necessary for a sucessful record linkage starting with data preparation  including pre-processing, cleaning and standardization of data.

## The Principles of Record Linkage
The goal of record linkage is to determine if pairs of records describe the same entity. For instance, this is important for removing duplicates from a data source or joining two separate data sources together. Record linkage also goes by the terms data matching, merge/purge, duplication detection, de-duping, reference matching, entity resolution, disambiguation, co-reference/anaphora in various fields.

There are several approaches to record linkage that include 
    - exact matching,
    - rule-based linking, 
    - probabilistic linking. 
- An example of **exact matching** is joining records based on a direct identifier. This is what we have already have done in SQL by joining tables. 
- **Rule-based matching** involves applying a cascading set of rules that reflect the domain knowledge of the records being linked. 
- In **probabilistic record linkages**, linkage weights are estimated to calculate the probability of a certain match.

In practical applications you will need record linkage techiques to combine information addressing the same entity that is stored in different data sources. Record linkage will also help you to address the quality of different data sources. For example, if one of your databases has missing values you might be able to fill those by finding an identical pair in a different data source. Overall, the main applications of record linkage are:
    1. Merging two or more data files. 
    2. Identifying the intersection of the two data sets. 
    3. Updating data files (with the data row of the other data files) and imputing missing data.
    4. Entity disambiguation and de-duplication.

## Linking Patents to Grants: individual and organization levels

In this class we will link records obtained from the **PatentsView** (https://www.patentsview.org/download/) to the data on federal grants in **FedRePORTER** (https://federalreporter.nih.gov/FileDownload). In both datasets we have names that we can use to link: inventor names on patents and principal investigator names on grants.

**About USPTO/PatentsView**

The PatentsView platform links inventors, organizations, locations, and patenting activity since 1976. 


**Key variables of interest**:

**Patent**
- Patent processing time (duration between a patent application filing date and
grant date)
- Title
- Type of patent document (utility, design, plant, reissue, defensive, and statutory
invention registration—offer different kinds of protection for distinct types of
subject matter)
- Cited patent (patent is cited when it appears as a reference in another patent or
patent application)
- Citing patent (patent that cites a patent or patent application as a reference in its
&#39;references cited&#39; section)
- Grant date (date a patent was issued by the USPTO)
- International Patent Classification (IPC) (system of patent classifications
published by the World Intellectual Property Organization (WIPO))

**Assignee** (the name of the entity—company, foundation, partnership, holding company
or individual—that owns the patent)
- First name (if assignee is individual)
- Last name (if assignee is individual)
- Organization name (if assignee is organization)
- Type (classification of assignee)
- Location (City, State, Country, Latitude, Longitude, County, FIPS code (state), FIPS code (county))

**Inventor** (the individual who conceived the invention)
- First name
- Last name
- Gender
- Location (City, State, Country, Latitude, Longitude, County, FIPS code (state), FIPS code (county))
- Co-inventor (a person who appears on the same patent or patent application as
another inventor)

**About Federal RePORTER**

Federal RePORTER is a collaborative effort led by STAR METRICS® to create a searchable
database of scientific awards from agencies (across agencies or fiscal years, by the award&#39;s
project leader, or by a text search of a project&#39;s title, terms, or abstracts). The National Institutes of Health (NIH) and the National Science Foundation (NSF), under the auspices of Office of Science and Technology Policy (OSTP), are leading this project with funding provided by NIH, NSF, the U.S. Department of Agriculture, and the Environmental Protection Agency.

Agencies with data in Federal RePORTER: DOD (Department of Defense), ED (Department of Education), Environmental Protection Agency (EPA), HHS (Department of Health and Human Services), National Aeronautics and Space Administration (NASA), National Science Foundation (NSF), USDA (United States Department of Agriculture), U.S. Department of Veterans Affairs (VA).

**Key variables of interest**:
- Project description
    - Title
    - Abstract
    - Terms (keywords from the project title and abstract)
- Project duration
    - Start date
    - End date
- Total cost of a project
- Financial year
- Federal agency, department
- Principal Investigator (PI) Project Leader
    - First name
    - Last name
- Other PIs
    - First name
    - Last name
- Organization name
- Organization location (City, State, Zip, Country)
- CFDA code (The Catalog of Federal Domestic Assistance, federal financial assistance
and nonfinancial assistance programs)

**Questions for group discussion:** 
- What variables would you use for record linkage using these datasets?
- What problems do you anticipate?

## Python Setup

**Class Subset**: The FedRePORTER dataset includes records from year 2010, and PatentsView dataset includes records from 2015-2019 of patents with government interest (in order to account for a time lag between the start of a grant and potential outputs (registered patents) from a grant).

Import necessary Python packages.

In [2]:
# data import and manipulation
import pandas as pd

# record linkage and preprocessing
import recordlinkage as rl
from recordlinkage.preprocessing import clean, phonetic

# Data Exploration: Know Your Data

### Individual-level links: inventor (patent) <-> principal investigator (grant)

#### PatentsView: inventor names on patents

In [3]:
# Import CSV into a dataframe
inventors = pd.read_csv('../Data/patentsview_inventors.csv')

In [6]:
# View first rows of the dataframe
inventors.head(10)

patent_id  year inventor_id       name_first       name_last inventor_city  \
0   9980635  2018   9980635-1     Maria Eliana   Manquez Hatta      Santiago   
1   9980635  2018   9980635-2         Humberto  Lobos Sucarrat      Santiago   
2   9999665  2018   8398993-2         Pablo A.  Gonzalez Munoz      Santiago   
3   9980641  2018   5321501-6          Joel S.         Schuman    Pittsburgh   
4   9980641  2018   9980641-3  Lawrence Edward        Kagemann    Pittsburgh   
5   9980645  2018   9037248-2            Brian       Wodlinger    Pittsburgh   
6   9981048  2018   9981048-1        Ira Jacob             Fox    Pittsburgh   
7   9981048  2018   9981048-3      Taichiro N.       Nishikawa    Pittsburgh   
8   9981048  2018   9981048-4        Alejandro  Soto-Gutierrez    Pittsburgh   
9   9981915  2018   5156079-2              Wei              Li    Pittsburgh   

  inventor_state inventor_country  
0            NaN               CL  
1            NaN               CL  
2            NaN               CL  
3             PA               US  
4             PA               US  
5             PA               US  
6             PA               US  
7             PA               US  
8             PA               US  
9             PA               US

In [8]:
inventors.tail(10)

patent_id  year inventor_id     name_first  name_last  \
4831   RE47009  2018   6534277-1       Wayne W.    Hancock   
4832   RE47026  2018   5805332-1          Anand   Gopinath   
4833   RE47026  2018   5805332-1          Anand   Gopinath   
4834   RE47026  2018   6650116-3          Lance  DelaBarre   
4835   RE47026  2018   6969992-4           John     Strupp   
4836   RE47026  2018   6969992-4           John     Strupp   
4837   RE47026  2018   6969992-5       Peter M.   Andersen   
4838   RE47026  2018   7800368-5  Heather Diane      Orser   
4839   RE47026  2018   RE47026-4        Chris E      Olson   
4840   RE47267  2019   5494810-2        John A.     Zebala   

            inventor_city inventor_state inventor_country  
4831             Medfield             MA               US  
4832              Wayzata             MN               US  
4833                Orono             MN               US  
4834          St. Anthony             MN               US  
4835  Marine on St. Croix             MN               US  
4836          Coon Rapids             MN               US  
4837         Independence             MN               US  
4838           Farmington             MN               US  
4839             Palatine             IL               US  
4840               Auburn             WA               US

In [7]:
inventors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4841 entries, 0 to 4840
Data columns (total 8 columns):
patent_id           4841 non-null object
year                4841 non-null int64
inventor_id         4841 non-null object
name_first          4841 non-null object
name_last           4841 non-null object
inventor_city       4840 non-null object
inventor_state      3994 non-null object
inventor_country    4841 non-null object
dtypes: int64(1), object(7)
memory usage: 302.6+ KB


We notice that `inventor_state` information appears to be missing in some cases (3,994 records vs total 4,841 records).

We can check the rows that are missing by calling `.isnull()`.

In [9]:
# Show rows which are null in `inventor_state` column
inventors[inventors['inventor_state'].isnull()]

patent_id  year inventor_id       name_first       name_last  \
0      9980635  2018   9980635-1     Maria Eliana   Manquez Hatta   
1      9980635  2018   9980635-2         Humberto  Lobos Sucarrat   
2      9999665  2018   8398993-2         Pablo A.  Gonzalez Munoz   
100    9980645  2018   4306291-3     Dominique M.          Durand   
101    9981915  2018   5156079-2              Wei              Li   
102    9983412  2018   D419566-1      David Scott            Dunn   
103    9984283  2018   7873169-1               Ke            Zeng   
104    9989702  2018   7826509-2      Benjamin G.             Lee   
105    9993465  2018   5326659-1             Yong             Liu   
106    9997705  2018   5546889-4             Yang            Yang   
107    9999704  2018   5081281-4             Jing           Zhang   
108    PP27141  2016   4395616-1         David C.           Smith   
109    PP28200  2017   4395616-1         David C.           Smith   
110    PP28216  2017   4395616-1         David C.           Smith   
111    RE46817  2018   7135040-1              Shu            Wang   
112    RE47157  2018   7421171-8         Rafif E.           Hamam   
127    9980645  2018   4306291-3     Dominique M.          Durand   
134    9980660  2018   5593835-2             Yong            Wang   
135    9986724  2018   4346055-1        Andrew J.          Murphy   
136    9987331  2018   7309693-4         Yasuhiro           Ikeda   
137    9991360  2018   4108044-1          Richard           Brown   
138    9994555  2018   6306837-2             Chao            Wang   
139    RE46953  2018   6306837-2             Chao            Wang   
140    RE46412  2017   7920271-2  Adrien Emmanuel      Desjardins   
229    9980660  2018   5593835-2             Yong            Wang   
230    9981338  2018   5418611-2              Jun              Ni   
231    9981915  2018   5156079-2              Wei              Li   
232    9981961  2018   5979076-1            Zheng              Li   
233    9982356  2018   5180707-1           Wei Z.             Gao   
234    9983285  2018   7583082-1             Peng              Hu   
...        ...   ...         ...              ...             ...   
4542   9999704  2018   5081281-4             Jing           Zhang   
4543   9999704  2018   5081281-4             Jing           Zhang   
4544   RE46129  2016   5247155-2              Lin              Li   
4545   9999704  2018   5081281-4             Jing           Zhang   
4546   9999704  2018   5081281-4             Jing           Zhang   
4549   PP25271  2015   8889951-1    Michael David         Materne   
4550   PP25272  2015   8889951-1    Michael David         Materne   
4551   PP25273  2015   8889951-1    Michael David         Materne   
4657   PP27141  2016   4395616-1         David C.           Smith   
4658   PP28200  2017   4395616-1         David C.           Smith   
4659   PP28216  2017   4395616-1         David C.           Smith   
4757   RE45512  2015   6831781-5   Milen Stefanov        Shishkov   
4779   RE46088  2016   5948958-2          Chen-An            Chen   
4780   RE46088  2016   8221601-2    Emmanuel Chua            Abas   
4781   RE46088  2016   8221601-2    Emmanuel Chua            Abas   
4782   RE46088  2016   8221601-2    Emmanuel Chua            Abas   
4783   RE46088  2016   8221601-4      Jake Randal          Ermita   
4784   RE46088  2016   8221601-5   Jose Francisco        Capulong   
4785   RE46088  2016   8221601-6  Arnold Villamor        Castillo   
4786   RE46088  2016   8221601-3    Edmundo Anida          Divino   
4787   RE46088  2016   8221601-5   Jose Francisco        Capulong   
4792   RE46129  2016   5247155-2              Lin              Li   
4794   RE46129  2016   5247155-2              Lin              Li   
4795   RE46129  2016   5247155-2              Lin              Li   
4796   RE46129  2016   5247155-2              Lin              Li   
4799   RE46129  2016   5498696-4         Xiaodong            Wang   
480

It looks like our records are not filtered by only US.

In [10]:
# Check unique values in the `inventor_country` column
inventors['inventor_country'].unique()

array(['CL', 'US', 'CA', 'FR', 'GB', 'CN', 'JP', 'KR', 'SG', 'DE', 'NL',
       'IN', 'CH', 'AU', 'TW', 'HK', 'FI', 'BE', 'SE', 'SW', 'EN', 'NM',
       'VN', 'BR', 'IE', 'SA', 'ES', 'IT', 'CZ', 'IL', 'RU', 'MX', 'AT',
       'DK', 'MO', 'UY', 'BG', 'PH'], dtype=object)

Let's filter by only those records which are in the US.

In [11]:
# Filter by those rows where country is US
inventors = inventors[inventors['inventor_country'] == 'US']

In [12]:
# Check that now we only have US as an inventor country
inventors['inventor_country'].unique()

array(['US'], dtype=object)

In [13]:
inventors['patent_id'].nunique()

593

We will be linking on name and location. What do you observe in the data?

In [14]:
# Check the tail of the table 
inventors.tail(10)

patent_id  year inventor_id     name_first  name_last  \
4831   RE47009  2018   6534277-1       Wayne W.    Hancock   
4832   RE47026  2018   5805332-1          Anand   Gopinath   
4833   RE47026  2018   5805332-1          Anand   Gopinath   
4834   RE47026  2018   6650116-3          Lance  DelaBarre   
4835   RE47026  2018   6969992-4           John     Strupp   
4836   RE47026  2018   6969992-4           John     Strupp   
4837   RE47026  2018   6969992-5       Peter M.   Andersen   
4838   RE47026  2018   7800368-5  Heather Diane      Orser   
4839   RE47026  2018   RE47026-4        Chris E      Olson   
4840   RE47267  2019   5494810-2        John A.     Zebala   

            inventor_city inventor_state inventor_country  
4831             Medfield             MA               US  
4832              Wayzata             MN               US  
4833                Orono             MN               US  
4834          St. Anthony             MN               US  
4835  Marine on St. Croix             MN               US  
4836          Coon Rapids             MN               US  
4837         Independence             MN               US  
4838           Farmington             MN               US  
4839             Palatine             IL               US  
4840               Auburn             WA               US

- First name includes first and middle names (potentially parse out)
- First and last names are capitalized
- The same person (with the same inventor id) can have different associated locations 

Check most frequent counts of variables.

In [15]:
inventors['inventor_city'].value_counts().head(10)

Cambridge        80
Los Angeles      60
San Diego        57
New York         48
Albuquerque      45
San Jose         44
Palo Alto        43
Baltimore        38
Pasadena         36
Mountain View    35
Name: inventor_city, dtype: int64

In [16]:
inventors['inventor_state'].value_counts().head(10)

CA    917
MA    400
NY    258
MD    176
FL    164
IL    163
PA    154
TX    142
NJ    129
CT    122
Name: inventor_state, dtype: int64

#### FedRePORTER: principal investigator names on grants

In [17]:
# Read-in the dataset into a dataframe
pi = pd.read_csv('../Data/fedreporter_pi_names_grants.csv')

In [18]:
# Show first rows of the dataframe
pi.head()

PROJECT_ID  CONTACT_PI_PROJECT_LEADER        OTHER_PIS  ORGANIZATION_CITY  \
0      121219              POIRER, DAVID              NaN             TUCSON   
1      121220             HEYWARD, ANN O              NaN         BROOK PARK   
2      121221        KALOGERA, VASSILIKI    LUIJTEN, ERIK           EVANSTON   
3      121222       GUILDERSON, THOMAS P              NaN         SANTA CRUZ   
4      121223             HAYES, GILLIAN  TENTORI, MONICA             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY  
0                  AZ         UNITED STATES  
1                  OH         UNITED STATES  
2                  IL         UNITED STATES  
3                  CA         UNITED STATES  
4                  CA         UNITED STATES

In [19]:
pi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111400 entries, 0 to 111399
Data columns (total 6 columns):
PROJECT_ID                    111400 non-null int64
 CONTACT_PI_PROJECT_LEADER    111398 non-null object
 OTHER_PIS                    10477 non-null object
 ORGANIZATION_CITY            110853 non-null object
 ORGANIZATION_STATE           109933 non-null object
 ORGANIZATION_COUNTRY         110861 non-null object
dtypes: int64(1), object(5)
memory usage: 5.1+ MB


Let's check which countries are there in the dataset.

In [20]:
# Call .unique() function
pi[' ORGANIZATION_COUNTRY'].unique()

array(['UNITED STATES', 'BRAZIL', 'BERMUDA', 'SWITZERLAND', 'FRANCE',
       'CANADA', 'UNITED KINGDOM', 'VENEZUELA', 'SOUTH AFRICA',
       'NEW ZEALAND', 'DENMARK', nan, 'SWEDEN', 'INDIA', 'SPAIN',
       'AUSTRALIA', 'PERU', 'HAITI', 'DOMINICAN REP', 'JAMAICA',
       'NIGERIA', 'ITALY', 'IRELAND', 'BELGIUM', 'CHINA', 'THAILAND',
       'JAPAN', 'ZAMBIA', 'ISRAEL', 'GERMANY', 'NETHERLANDS', 'UGANDA',
       'PAKISTAN', 'KOREA REP OF', 'CHILE', 'TANZANIA U REP', 'ARGENTINA',
       'HUNGARY', 'URUGUAY', 'RUSSIA', 'KENYA', 'MALAWI', 'LITHUANIA',
       'FINLAND', 'MEXICO', 'COLOMBIA', 'CAMBODIA', 'PHILIPPINES',
       'TURKEY', 'ZIMBABWE', 'BOTSWANA', 'CROATIA', 'COSTA RICA',
       'TUNISIA', 'ICELAND', 'MOZAMBIQUE', 'GHANA', 'SINGAPORE',
       'LEBANON', 'PARAGUAY', 'ETHIOPIA', 'UKRAINE', 'LESOTHO',
       "COTE D'IVOIRE", 'DOMINICA', 'NAMIBIA', 'GUATEMALA', 'RWANDA',
       'GEORGIA', 'VIETNAM', 'MOLDOVA', 'TRINIDAD/TOBA', 'CAMEROON',
       'ANGOLA', 'INDONESIA', 'CONGO', 'CONGO 

Let's filter by only US like we did with the PatentsView dataset.

In [21]:
# Filter by only those rows with US as an organization_country
pi = pi[pi[' ORGANIZATION_COUNTRY'] == 'UNITED STATES']

In [22]:
# Call .unique() function
pi[' ORGANIZATION_COUNTRY'].unique()

array(['UNITED STATES'], dtype=object)

In [23]:
pi[' ORGANIZATION_CITY'].value_counts().head(10)

BOSTON          5802
NEW YORK        4603
BEHTESDA        3745
PHILADELPHIA    3044
BALTIMORE       2957
SEATTLE         2917
LOS ANGELES     2905
LA JOLLA        2835
CHICAGO         2523
HOUSTON         2345
Name:  ORGANIZATION_CITY, dtype: int64

In [24]:
pi[' ORGANIZATION_STATE'].value_counts().head(10)

CA    15293
NY     9186
MA     8812
MD     7932
PA     6245
TX     5653
NC     4036
IL     3796
WA     3409
OH     3034
Name:  ORGANIZATION_STATE, dtype: int64

**Compare inventor and grants table: what types of cleaning/standardization would we need to do?**

In [25]:
inventors.head()

patent_id  year inventor_id       name_first  name_last inventor_city  \
3   9980641  2018   5321501-6          Joel S.    Schuman    Pittsburgh   
4   9980641  2018   9980641-3  Lawrence Edward   Kagemann    Pittsburgh   
5   9980645  2018   9037248-2            Brian  Wodlinger    Pittsburgh   
6   9981048  2018   9981048-1        Ira Jacob        Fox    Pittsburgh   
7   9981048  2018   9981048-3      Taichiro N.  Nishikawa    Pittsburgh   

  inventor_state inventor_country  
3             PA               US  
4             PA               US  
5             PA               US  
6             PA               US  
7             PA               US

In [26]:
pi.head()

PROJECT_ID  CONTACT_PI_PROJECT_LEADER        OTHER_PIS  ORGANIZATION_CITY  \
0      121219              POIRER, DAVID              NaN             TUCSON   
1      121220             HEYWARD, ANN O              NaN         BROOK PARK   
2      121221        KALOGERA, VASSILIKI    LUIJTEN, ERIK           EVANSTON   
3      121222       GUILDERSON, THOMAS P              NaN         SANTA CRUZ   
4      121223             HAYES, GILLIAN  TENTORI, MONICA             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY  
0                  AZ         UNITED STATES  
1                  OH         UNITED STATES  
2                  IL         UNITED STATES  
3                  CA         UNITED STATES  
4                  CA         UNITED STATES

<span style="color:red">**Checkpoint 1: Explore organization-level data from PatentsView and FedRePORTER**</span>

Now try exploring organization-level data for both PatentsView and FedRePORTER using functions for data exploration used in the examples above.

In [27]:
assignees = pd.read_csv('../Data/patentsview_assignee_org_names.csv')

In [28]:
organizations = pd.read_csv('../Data/fedreporter_org_names_grants.csv')

In [31]:
#Check assignee data from patentview
assignees.head()

patent_id                   organization   org_city org_state org_country
0   9992413               Raytheon Company  Lexington        MA          US
1   RE45718  Boston Scientific Corporation  Lexington        MA          US
2   9992413               Raytheon Company    Waltham        MA          US
3   9988932   Honeywell International Inc.    Waltham        MA          US
4   9991578   Honeywell International Inc.    Waltham        MA          US

In [33]:
assignees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4371 entries, 0 to 4370
Data columns (total 5 columns):
patent_id       4371 non-null object
organization    4371 non-null object
org_city        4224 non-null object
org_state       3869 non-null object
org_country     4371 non-null object
dtypes: object(5)
memory usage: 170.8+ KB


In [34]:
assignees['org_country'].unique()

array(['US', 'JP', 'IL', 'KR', 'CA', 'TW', 'FI', 'NL', 'DE', 'DT', 'SC',
       'AU', 'GB', 'MX', 'CH', 'FR', 'AN', 'CZ', 'DK', 'SE', 'EN', 'un',
       'PR', 'CN', 'BM', 'BB', 'KN', 'CL'], dtype=object)

In [36]:
# Filter by only those rows with US as an organization_country
assignees = assignees[assignees['org_country'] == 'US']
assignees['org_country'].unique()

array(['US'], dtype=object)

In [37]:
#Check organization data from FedReporters
organizations.head()

PROJECT_ID                    ORGANIZATION_NAME  ORGANIZATION_CITY  \
0      121219                UNIVERSITY OF ARIZONA             TUCSON   
1      121220             OHIO AEROSPACE INSTITUTE         BROOK PARK   
2      121221              NORTHWESTERN UNIVERSITY           EVANSTON   
3      121222  UNIVERSITY OF CALIFORNIA SANTA CRUZ         SANTA CRUZ   
4      121223      UNIVERSITY OF CALIFORNIA IRVINE             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY  
0                  AZ         UNITED STATES  
1                  OH         UNITED STATES  
2                  IL         UNITED STATES  
3                  CA         UNITED STATES  
4                  CA         UNITED STATES

In [38]:
organizations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111400 entries, 0 to 111399
Data columns (total 5 columns):
PROJECT_ID               111400 non-null int64
 ORGANIZATION_NAME       111203 non-null object
 ORGANIZATION_CITY       110853 non-null object
 ORGANIZATION_STATE      109933 non-null object
 ORGANIZATION_COUNTRY    110861 non-null object
dtypes: int64(1), object(4)
memory usage: 4.2+ MB


In [41]:
organizations[' ORGANIZATION_COUNTRY'].unique()

array(['UNITED STATES', 'BRAZIL', 'BERMUDA', 'SWITZERLAND', 'FRANCE',
       'CANADA', 'UNITED KINGDOM', 'VENEZUELA', 'SOUTH AFRICA',
       'NEW ZEALAND', 'DENMARK', nan, 'SWEDEN', 'INDIA', 'SPAIN',
       'AUSTRALIA', 'PERU', 'HAITI', 'DOMINICAN REP', 'JAMAICA',
       'NIGERIA', 'ITALY', 'IRELAND', 'BELGIUM', 'CHINA', 'THAILAND',
       'JAPAN', 'ZAMBIA', 'ISRAEL', 'GERMANY', 'NETHERLANDS', 'UGANDA',
       'PAKISTAN', 'KOREA REP OF', 'CHILE', 'TANZANIA U REP', 'ARGENTINA',
       'HUNGARY', 'URUGUAY', 'RUSSIA', 'KENYA', 'MALAWI', 'LITHUANIA',
       'FINLAND', 'MEXICO', 'COLOMBIA', 'CAMBODIA', 'PHILIPPINES',
       'TURKEY', 'ZIMBABWE', 'BOTSWANA', 'CROATIA', 'COSTA RICA',
       'TUNISIA', 'ICELAND', 'MOZAMBIQUE', 'GHANA', 'SINGAPORE',
       'LEBANON', 'PARAGUAY', 'ETHIOPIA', 'UKRAINE', 'LESOTHO',
       "COTE D'IVOIRE", 'DOMINICA', 'NAMIBIA', 'GUATEMALA', 'RWANDA',
       'GEORGIA', 'VIETNAM', 'MOLDOVA', 'TRINIDAD/TOBA', 'CAMEROON',
       'ANGOLA', 'INDONESIA', 'CONGO', 'CONGO 

In [43]:
organizations = organizations[organizations[' ORGANIZATION_COUNTRY'] == 'UNITED STATES']
organizations[' ORGANIZATION_COUNTRY'].unique()

array(['UNITED STATES'], dtype=object)

In [45]:
assignees.head()


patent_id                   organization   org_city org_state org_country
0   9992413               Raytheon Company  Lexington        MA          US
1   RE45718  Boston Scientific Corporation  Lexington        MA          US
2   9992413               Raytheon Company    Waltham        MA          US
3   9988932   Honeywell International Inc.    Waltham        MA          US
4   9991578   Honeywell International Inc.    Waltham        MA          US

In [46]:
organizations.head()

PROJECT_ID                    ORGANIZATION_NAME  ORGANIZATION_CITY  \
0      121219                UNIVERSITY OF ARIZONA             TUCSON   
1      121220             OHIO AEROSPACE INSTITUTE         BROOK PARK   
2      121221              NORTHWESTERN UNIVERSITY           EVANSTON   
3      121222  UNIVERSITY OF CALIFORNIA SANTA CRUZ         SANTA CRUZ   
4      121223      UNIVERSITY OF CALIFORNIA IRVINE             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY  
0                  AZ         UNITED STATES  
1                  OH         UNITED STATES  
2                  IL         UNITED STATES  
3                  CA         UNITED STATES  
4                  CA         UNITED STATES

In [49]:
assignees['patent_id'].nunique()

582

In [51]:
organizations['PROJECT_ID'].nunique()

109746

# Data Cleaning

## The Importance of Pre-Processing
Data pre-processing is an important step in a data anlysis project in general, in record linkage applications in particular. The goal of pre-processing is to transform messy data into a dataset that can be used in a project workflow.

Linking records from different data sources comes with different challenges that need to be addressed by the analyst. The analyst must determine whether or not two entities (individuals, businesses, geographical units) on two different files are the same. This determination is not always easy. In most of the cases there is no common uniquely identifing characteristic for a entity. For example, is Bob Miller from New Yor the same person as Bob Miller from Chicago in a given dataset? This detemination has to be executed carefully because consequences of wrong linkages may be substantial (is person X the same person as the person X on the list of identified terrorists). Pre-processing can help to make better informed decisions.

Pre-processing can be difficult because there are a lot of things to keep in mind. For example, data input errors, such as typos, misspellings, truncation, abbreviations, and missing values need to be corrected. Literature shows that preprocessing can improve matches. 

“In situations of reasonably high-quality data, preprocessing can yield a greater improvement in matching efficiency than string comparators and ‘optimized’ parameters. In some situations, 90% of the improvement in matching efficiency may be due to preprocessing” (Winkler 2009).

The most common reason why matching projects fail is lack of time and resources for data cleaning. 

In the following section we will walk you through some pre-processing steps, these include but are not limited to removing spaces, parsing fields, and standardizing strings.

### Individual-level links

#### Clean Inventor Data
We will start by cleaning and preprocessing the inventor data. We need to remove whitespaces, make sure that everything is in lower case, we need to parse the first name and the last name, and harmonize all the other information we need for the linkage.

The `recordlinkage` package comes with a built-in cleaning function we can use. The `clean()` function removes any characters such as `-`, `.`, `/`, `:`, brackets of all types, and also lowercases by default.

In [52]:
# Let's take a look at names before
inventors.head()

patent_id  year inventor_id       name_first  name_last inventor_city  \
3   9980641  2018   5321501-6          Joel S.    Schuman    Pittsburgh   
4   9980641  2018   9980641-3  Lawrence Edward   Kagemann    Pittsburgh   
5   9980645  2018   9037248-2            Brian  Wodlinger    Pittsburgh   
6   9981048  2018   9981048-1        Ira Jacob        Fox    Pittsburgh   
7   9981048  2018   9981048-3      Taichiro N.  Nishikawa    Pittsburgh   

  inventor_state inventor_country  
3             PA               US  
4             PA               US  
5             PA               US  
6             PA               US  
7             PA               US

In [53]:
# Cleaning names (using the record linkage package tool)
inventors['name_last'] = clean(inventors['name_last'])
inventors['name_first'] = clean(inventors['name_first'])

In [54]:
# Let's take a look at names after the cleaning
inventors.head()

patent_id  year inventor_id       name_first  name_last inventor_city  \
3   9980641  2018   5321501-6           joel s    schuman    Pittsburgh   
4   9980641  2018   9980641-3  lawrence edward   kagemann    Pittsburgh   
5   9980645  2018   9037248-2            brian  wodlinger    Pittsburgh   
6   9981048  2018   9981048-1        ira jacob        fox    Pittsburgh   
7   9981048  2018   9981048-3       taichiro n  nishikawa    Pittsburgh   

  inventor_state inventor_country  
3             PA               US  
4             PA               US  
5             PA               US  
6             PA               US  
7             PA               US

We can parse and extract middle names.

In [55]:
# Split by white space and extract first element for the first name,
# second element for the middle name
inventors['name_middle'] = inventors.name_first.str.split(' ').str.get(1)
inventors['name_first'] = inventors.name_first.str.split(' ').str.get(0)

In [56]:
inventors.head()

patent_id  year inventor_id name_first  name_last inventor_city  \
3   9980641  2018   5321501-6       joel    schuman    Pittsburgh   
4   9980641  2018   9980641-3   lawrence   kagemann    Pittsburgh   
5   9980645  2018   9037248-2      brian  wodlinger    Pittsburgh   
6   9981048  2018   9981048-1        ira        fox    Pittsburgh   
7   9981048  2018   9981048-3   taichiro  nishikawa    Pittsburgh   

  inventor_state inventor_country name_middle  
3             PA               US           s  
4             PA               US      edward  
5             PA               US         NaN  
6             PA               US       jacob  
7             PA               US           n

Now we are done with the inital data prep work for the inventor file. Please keep in mind that we just provided some examples for you to demonstrate the process. You can add as many further steps to it as necessary. 

#### Clean Grant Data
Now we will clean and preprocess the grants data. We need a dataset that has the first, middle and last name of the researcher, the city, state, and country information. 

In [57]:
pi.head()

PROJECT_ID  CONTACT_PI_PROJECT_LEADER        OTHER_PIS  ORGANIZATION_CITY  \
0      121219              POIRER, DAVID              NaN             TUCSON   
1      121220             HEYWARD, ANN O              NaN         BROOK PARK   
2      121221        KALOGERA, VASSILIKI    LUIJTEN, ERIK           EVANSTON   
3      121222       GUILDERSON, THOMAS P              NaN         SANTA CRUZ   
4      121223             HAYES, GILLIAN  TENTORI, MONICA             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY  
0                  AZ         UNITED STATES  
1                  OH         UNITED STATES  
2                  IL         UNITED STATES  
3                  CA         UNITED STATES  
4                  CA         UNITED STATES

Here you can start cleaning the data and extract the information that we need to compare it to the inventor file. You want to find as many variables as you can match to as possible. You can also see that we have two field with names. You have the PI field, but you can extract more names from the OTHER PIDS fields if you are interested. 

In [58]:
pi[' CONTACT_PI_PROJECT_LEADER'] = clean(pi[' CONTACT_PI_PROJECT_LEADER'])

In [59]:
pi['name_first'] = pi[' CONTACT_PI_PROJECT_LEADER'].str.split(' ').str.get(1)
pi['name_last'] = pi[' CONTACT_PI_PROJECT_LEADER'].str.split(' ').str.get(0)
pi['name_middle'] = pi[' CONTACT_PI_PROJECT_LEADER'].str.split(' ').str.get(2)

In [60]:
pi.head()

PROJECT_ID  CONTACT_PI_PROJECT_LEADER        OTHER_PIS  ORGANIZATION_CITY  \
0      121219               poirer david              NaN             TUCSON   
1      121220              heyward ann o              NaN         BROOK PARK   
2      121221         kalogera vassiliki    LUIJTEN, ERIK           EVANSTON   
3      121222        guilderson thomas p              NaN         SANTA CRUZ   
4      121223              hayes gillian  TENTORI, MONICA             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY name_first   name_last name_middle  
0                  AZ         UNITED STATES      david      poirer         NaN  
1                  OH         UNITED STATES        ann     heyward           o  
2                  IL         UNITED STATES  vassiliki    kalogera         NaN  
3                  CA         UNITED STATES     thomas  guilderson           p  
4                  CA         UNITED STATES    gillian       hayes         NaN

We will add one more thing: **phonetic processing**.

Sometimes, words or names are recorded differently because they are written down as they sound. This can result in failed matches, because the same organization or individual will technically have different written names, even though the names would sound identically when pronounced out loud. To avoid these issues, we will add onoe more thing: a soundex (a phonetic algorithm for indexing names by sound, as pronounced in English).

The `phonetic()` function is used to convert strings into their corresponding phonetic codes. This is particularly useful when comparing names where different possible spellings make it difficult to find exact matches (e.g. Jillian and Gillian).

Let's add a column called `phonetic_name` to our existing table, which will contain the result of applying a `phonetic` function to the name (the phonetic transcription of the name). We are using a method called NYSIIS - the New York State Identification and Intelligence System phonetic code.

In [61]:
inventors["phonetic_first"] = phonetic(inventors["name_first"], method="nysiis")
inventors["phonetic_last"] = phonetic(inventors["name_last"], method="nysiis")
inventors["phonetic_middle"] = phonetic(inventors["name_middle"], method="nysiis")

pi["phonetic_first"] = phonetic(pi["name_first"], method="nysiis")
pi["phonetic_last"] = phonetic(pi["name_last"], method="nysiis")
pi["phonetic_middle"] = phonetic(pi["name_middle"], method="nysiis")

In [62]:
inventors.head()

patent_id  year inventor_id name_first  name_last inventor_city  \
3   9980641  2018   5321501-6       joel    schuman    Pittsburgh   
4   9980641  2018   9980641-3   lawrence   kagemann    Pittsburgh   
5   9980645  2018   9037248-2      brian  wodlinger    Pittsburgh   
6   9981048  2018   9981048-1        ira        fox    Pittsburgh   
7   9981048  2018   9981048-3   taichiro  nishikawa    Pittsburgh   

  inventor_state inventor_country name_middle phonetic_first phonetic_last  \
3             PA               US           s            JAL         SANAN   
4             PA               US      edward         LARANC       CAGANAN   
5             PA               US         NaN           BRAN     WADLANGAR   
6             PA               US       jacob             IR           FAX   
7             PA               US           n          TACAR         NASAC   

  phonetic_middle  
3               S  
4           EDWAD  
5             NaN  
6           JACAB  
7               N

In [63]:
pi.head()

PROJECT_ID  CONTACT_PI_PROJECT_LEADER        OTHER_PIS  ORGANIZATION_CITY  \
0      121219               poirer david              NaN             TUCSON   
1      121220              heyward ann o              NaN         BROOK PARK   
2      121221         kalogera vassiliki    LUIJTEN, ERIK           EVANSTON   
3      121222        guilderson thomas p              NaN         SANTA CRUZ   
4      121223              hayes gillian  TENTORI, MONICA             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY name_first   name_last  \
0                  AZ         UNITED STATES      david      poirer   
1                  OH         UNITED STATES        ann     heyward   
2                  IL         UNITED STATES  vassiliki    kalogera   
3                  CA         UNITED STATES     thomas  guilderson   
4                  CA         UNITED STATES    gillian       hayes   

  name_middle phonetic_first phonetic_last phonetic_middle  
0         NaN          DAVAD         PARAR             NaN  
1           o             AN        HAYWAD               O  
2         NaN        VASALAC       CALAGAR             NaN  
3           p            TAN     GALDARSAN               P  
4         NaN          GALAN           HAY             NaN

<span style="color:red">**Checkpoint 2: Clean organization-level data**</span>

Use `clean` function above, remove white space, and find phonetic transcriptions of organization names.

In [64]:
# Clean assignee data
assignees.head()

patent_id                   organization   org_city org_state org_country
0   9992413               Raytheon Company  Lexington        MA          US
1   RE45718  Boston Scientific Corporation  Lexington        MA          US
2   9992413               Raytheon Company    Waltham        MA          US
3   9988932   Honeywell International Inc.    Waltham        MA          US
4   9991578   Honeywell International Inc.    Waltham        MA          US

In [67]:
# Cleaning names (using the record linkage package tool)
assignees['organization'] = clean(assignees['organization'])

In [71]:
assignees.head()

patent_id                   organization   org_city org_state org_country
0   9992413               raytheon company  Lexington        MA          US
1   RE45718  boston scientific corporation  Lexington        MA          US
2   9992413               raytheon company    Waltham        MA          US
3   9988932    honeywell international inc    Waltham        MA          US
4   9991578    honeywell international inc    Waltham        MA          US

In [73]:
assignees['organization'].unique()

array(['raytheon company', 'boston scientific corporation',
       'honeywell international inc', 'trustees of boston university',
       'brandeis university', 'boston dynamics inc',
       'california institute of technology',
       'university of southern california',
       'the regents of the university of california',
       'northrop grumman corporation',
       'international business machines corporation',
       'board of regents the university of texas system',
       'lockheed martin corporation',
       'the board of regents of the university of texas system',
       'the arizona board of regents on behalf of the university of arizona',
       'bat reseach and consulting',
       'arizona board of regents on behalf of arizona state university',
       'sion power corporation', 'paragon space development corporation',
       'general electric company', 'rolls royce plc',
       'consolidated nuclear security llc', 'leidos inc',
       'united technologies corporation', 'ya

In [76]:
assignees['organization']=assignees['organization'].str.replace('company','')

In [79]:
assignees['organization']=assignees['organization'].str.replace('corporation','')

In [80]:
assignees['organization']=assignees['organization'].str.replace('the ','')

In [81]:
assignees['organization']=assignees['organization'].str.replace(' llc','')

In [83]:
assignees['organization']=assignees['organization'].str.replace(' inc','')

In [84]:
assignees['organization'] = clean(assignees['organization'])

In [86]:
assignees['organization'].unique()

array(['raytheon', 'boston scientific', 'honeywell international',
       'trustees of boston university', 'brandeis university',
       'boston dynamics', 'california institute of technology',
       'university of southern california',
       'regents of university of california', 'northrop grumman',
       'international business machines',
       'board of regents university of texas system', 'lockheed martin',
       'board of regents of university of texas system',
       'arizona board of regents on behalf of university of arizona',
       'bat reseach and consulting',
       'arizona board of regents on behalf of arizona state university',
       'sion power', 'paragon space development', 'general electric',
       'rolls royce plc', 'consolidated nuclear security', 'leidos',
       'united technologies', 'yale university', 'hamilton sundstrand',
       'carrier', 'university of connecticut', 'goodrich',
       'university of connecticut technology commerialization services',
 

In [89]:
assignees['org_phonetic'] = phonetic(assignees['organization'], method="nysiis")
assignees.head(20)

patent_id                        organization     org_city org_state  \
0    9992413                            raytheon    Lexington        MA   
1    RE45718                   boston scientific    Lexington        MA   
2    9992413                            raytheon      Waltham        MA   
3    9988932             honeywell international      Waltham        MA   
4    9991578             honeywell international      Waltham        MA   
5    9989730             honeywell international      Waltham        MA   
6    9993577       trustees of boston university      Waltham        MA   
7    9986740                 brandeis university      Waltham        MA   
8    9982790                     boston dynamics      Waltham        MA   
9    9992413                            raytheon  Los Angeles        CA   
10   9986934  california institute of technology  Los Angeles        CA   
11   9995859  california institute of technology  Los Angeles        CA   
12   9988653  california institute of technology  Los Angeles        CA   
13   9991677  california institute of technology  Los Angeles        CA   
14   9983397  california institute of technology  Los Angeles        CA   
15   9998658  california institute of technology  Los Angeles        CA   
16   9981389  california institute of technology  Los Angeles        CA   
17   9993149  california institute of technology  Los Angeles        CA   
18   9989680  california institute of technology  Los Angeles        CA   
19   9995930  california institute of technology  Los Angeles        CA   

   org_country                 org_phonetic  
0           US                       RAYTAN  
1           US              BASTANSCANTAFAC  
2           US                       RAYTAN  
3           US         HANAYWALANTARNATANAL  
4           US         HANAYWALANTARNATANAL  
5           US         HANAYWALANTARNATANAL  
6           US    TRASTASAFBASTANANAVARSATY  
7           US            BRANDASANAVARSATY  
8           US                BASTANDYNANAC  
9           US                       RAYTAN  
10          US  CALAFARNANSTATATAFTACNALAGY  
11          US  CALAFARNANSTATATAFTACNALAGY  
12          US  CALAFARNANSTATATAFTACNALAGY  
13          US  CALAFARNANSTATATAFTACNALAGY  
14          US  CALAFARNANSTATATAFTACNALAGY  
15          US  CALAFARNANSTATATAFTACNALAGY  
16          US  CALAFARNANSTATATAFTACNALAGY  
17          US  CALAFARNANSTATATAFTACNALAGY  
18          US  CALAFARNANSTATATAFTACNALAGY  
19          US  CALAFARNANSTATATAFTACNALAGY

In [90]:
# Clean organization data
organizations.head()

PROJECT_ID                    ORGANIZATION_NAME  ORGANIZATION_CITY  \
0      121219                UNIVERSITY OF ARIZONA             TUCSON   
1      121220             OHIO AEROSPACE INSTITUTE         BROOK PARK   
2      121221              NORTHWESTERN UNIVERSITY           EVANSTON   
3      121222  UNIVERSITY OF CALIFORNIA SANTA CRUZ         SANTA CRUZ   
4      121223      UNIVERSITY OF CALIFORNIA IRVINE             IRVINE   

   ORGANIZATION_STATE  ORGANIZATION_COUNTRY  
0                  AZ         UNITED STATES  
1                  OH         UNITED STATES  
2                  IL         UNITED STATES  
3                  CA         UNITED STATES  
4                  CA         UNITED STATES

In [95]:
organizations[' ORGANIZATION_NAME'] = clean(organizations[' ORGANIZATION_NAME'])

In [100]:
organizations[' ORGANIZATION_NAME'].unique()

array(['university of arizona', 'ohio aerospace institute',
       'northwestern university', ..., 'stemina biomarker discovery inc',
       'achaogen inc', 'defense and veterans brain injury center'],
      dtype=object)

In [102]:
organizations[' ORGANIZATION_NAME']=organizations[' ORGANIZATION_NAME'].str.replace('company','')
organizations[' ORGANIZATION_NAME']=organizations[' ORGANIZATION_NAME'].str.replace('corporation','')
organizations[' ORGANIZATION_NAME']=organizations[' ORGANIZATION_NAME'].str.replace('the ','')
organizations[' ORGANIZATION_NAME']=organizations[' ORGANIZATION_NAME'].str.replace(' llc','')
organizations[' ORGANIZATION_NAME']=organizations[' ORGANIZATION_NAME'].str.replace(' inc','')
organizations[' ORGANIZATION_NAME'] = clean(organizations[' ORGANIZATION_NAME'])
organizations['org_phonetic'] = phonetic(organizations[' ORGANIZATION_NAME'], method="nysiis")
organizations.head(20)

PROJECT_ID                                 ORGANIZATION_NAME  \
0       121219                             university of arizona   
1       121220                          ohio aerospace institute   
2       121221                           northwestern university   
3       121222               university of california santa cruz   
4       121223                   university of california irvine   
5       121224                          georgia state university   
6       121225                  virginia commonwealth university   
7       121226                  montana state university bozeman   
8       121227                             mount holyoke college   
9       121228                    university of alaska fairbanks   
10      121229                             university of montana   
11      121230                university of massachusetts boston   
12      121231                           kansas state university   
13      121232                new jersey institute of technology   
14      121233                   consortium for ocean leadership   
15      121234                 rochester institute of technology   
16      121235                          educational broadcasting   
17      121236                             university of florida   
18      121237                university of california san diego   
19      121238  optoelectronics industry development association   

    ORGANIZATION_CITY  ORGANIZATION_STATE  ORGANIZATION_COUNTRY  \
0              TUCSON                  AZ         UNITED STATES   
1          BROOK PARK                  OH         UNITED STATES   
2            EVANSTON                  IL         UNITED STATES   
3          SANTA CRUZ                  CA         UNITED STATES   
4              IRVINE                  CA         UNITED STATES   
5             ATLANTA                  GA         UNITED STATES   
6            RICHMOND                  VA         UNITED STATES   
7             BOZEMAN                  MT         UNITED STATES   
8        SOUTH HADLEY                  MA         UNITED STATES   
9           FAIRBANKS                  AK         UNITED STATES   
10           MISSOULA                  MT         UNITED STATES   
11             LOWELL                  MA         UNITED STATES   
12          MANHATTAN                  KS         UNITED STATES   
13             NEWARK                  NJ         UNITED STATES   
14         WASHINGTON                  DC         UNITED STATES   
15          ROCHESTER                  NY         UNITED STATES   
16           NEW YORK                  NY         UNITED STATES   
17        GAINESVILLE                  FL         UNITED STATES   
18           LA JOLLA                  CA         UNITED STATES   
19         WASHINGTON                  DC         UNITED STATES   

                                 org_phonetic  
0                          UNAVARSATYAFARASAN  
1                           OHARASPACANSTATAT  
2                       NARTWASTARNANAVARSATY  
3                UNAVARSATYAFCALAFARNASANTACR  
4                   UNAVARSATYAFCALAFARNARVAN  
5                         GARGASTATANAVARSATY  
6                  VARGANACANANWALTANAVARSATY  
7                MANTANASTATANAVARSATYBASANAN  
8                             MANTALYACACALAG  
9                   UNAVARSATYAFALASCAFARBANC  
10                         UNAVARSATYAFNANTAN  
11               UNAVARSATYAFNASACASATSBASTAN  
12                        CANSASTATANAVARSATY  
13               NAEJARSAYANSTATATAFTACNALAGY  
14                   CANSARTANFARACANLADARSAP  
15                RACASTARANSTATATAFTACNALAGY  
16                      EDACATANALBRADCASTANG  
17                          UNAVARSATYAFLARAD  
18                UNAVARSATYAFCALAFARNASANDAG  
19  OPTALACTRANACSANDASTRYDAFALAPNANTASACATAN

In [106]:
#Merging
merge_org = assignees.merge(organizations,on=['org_phonetic'])

In [108]:
merge_org['patent_id'].nunique()

195

In [112]:
assignees['patent_id'].nunique()

582

In [111]:
#How many patents issued bwt 2015-2019 are from organizations that received grants in 2010
merge_org['patent_id'].nunique()/assignees['patent_id'].nunique()

0.33505154639175255

In [115]:
#How many granted projects in 2010 are from the organizations that gained patents bwt 2015-2019
merge_org['PROJECT_ID'].nunique()/organizations['PROJECT_ID'].nunique()

0.19348313378164125

# Deterministic Matching

### Individual-level links

Let's first do an exact match (merge) on the **first, middle, last name** and examine the results.

In [116]:
# Merge on first, middle, last name
merge_name = inventors.merge(pi,on=['name_first','name_last','name_middle'])

Let's count how many unique inventor IDs and PI names got matched.

In [117]:
# Let's check the count of unique inventor_ids
merge_name['inventor_id'].nunique()

250

We can now merge on the **phonetic first, middle, last name**.

In [118]:
# Merge on phonetic first, middle, last name
merge_name_phonetic_middle = inventors.merge(pi,on=['phonetic_first','phonetic_last','phonetic_middle'])

In [119]:
# Let's check the count of unique inventor_ids
merge_name_phonetic_middle['inventor_id'].nunique()

307

You can enlarge your links and only merge on **first and last name**.

In [120]:
merge_name_first_last = inventors.merge(pi,on=['name_first','name_last'])

In [121]:
# Check the counts
merge_name_first_last['inventor_id'].nunique()

384

You can also try merging on **phonetic first and last name**.

In [122]:
merge_name_phonetic = inventors.merge(pi,on=['phonetic_first','phonetic_last'])

In [123]:
merge_name_phonetic['inventor_id'].nunique()

537

What if we blocked by state?

In [124]:
# Rename columns in both datasets to the same name
inventors = inventors.rename(columns={'inventor_state':'state'})
pi = pi.rename(columns={' ORGANIZATION_STATE':'state'})

Merge on **first, middle, last name, and state**.

In [125]:
merge_name_state = inventors.merge(pi,on=['name_first','name_last','name_middle','state'])

How many matches did we get?

In [126]:
merge_name_state['inventor_id'].nunique()

221

In [127]:
merge_phonetic_name_state = inventors.merge(pi,on=['phonetic_first','phonetic_last','phonetic_middle','state'])

In [128]:
merge_phonetic_name_state['inventor_id'].nunique()

244

Now let's try blocking by **full name, city and state**.

In [129]:
inventors.head()

patent_id  year inventor_id name_first  name_last inventor_city state  \
3   9980641  2018   5321501-6       joel    schuman    Pittsburgh    PA   
4   9980641  2018   9980641-3   lawrence   kagemann    Pittsburgh    PA   
5   9980645  2018   9037248-2      brian  wodlinger    Pittsburgh    PA   
6   9981048  2018   9981048-1        ira        fox    Pittsburgh    PA   
7   9981048  2018   9981048-3   taichiro  nishikawa    Pittsburgh    PA   

  inventor_country name_middle phonetic_first phonetic_last phonetic_middle  
3               US           s            JAL         SANAN               S  
4               US      edward         LARANC       CAGANAN           EDWAD  
5               US         NaN           BRAN     WADLANGAR             NaN  
6               US       jacob             IR           FAX           JACAB  
7               US           n          TACAR         NASAC               N

In [130]:
pi.head()

PROJECT_ID  CONTACT_PI_PROJECT_LEADER        OTHER_PIS  ORGANIZATION_CITY  \
0      121219               poirer david              NaN             TUCSON   
1      121220              heyward ann o              NaN         BROOK PARK   
2      121221         kalogera vassiliki    LUIJTEN, ERIK           EVANSTON   
3      121222        guilderson thomas p              NaN         SANTA CRUZ   
4      121223              hayes gillian  TENTORI, MONICA             IRVINE   

  state  ORGANIZATION_COUNTRY name_first   name_last name_middle  \
0    AZ         UNITED STATES      david      poirer         NaN   
1    OH         UNITED STATES        ann     heyward           o   
2    IL         UNITED STATES  vassiliki    kalogera         NaN   
3    CA         UNITED STATES     thomas  guilderson           p   
4    CA         UNITED STATES    gillian       hayes         NaN   

  phonetic_first phonetic_last phonetic_middle  
0          DAVAD         PARAR             NaN  
1             AN        HAYWAD               O  
2        VASALAC       CALAGAR             NaN  
3            TAN     GALDARSAN               P  
4          GALAN           HAY             NaN

In [131]:
inventors['inventor_city'] = inventors['inventor_city'].str.upper()

In [132]:
inventors = inventors.rename(columns={'inventor_city':'city'})

In [133]:
pi = pi.rename(columns={' ORGANIZATION_CITY':'city'})

In [134]:
merge_name_location_city = inventors.merge(pi, on=['name_first','name_last','name_middle','state','city'])

How many matches did we get?

In [135]:
merge_name_location_city['inventor_id'].nunique()

135

In [136]:
merge_phonetic_name_location_city = inventors.merge(pi, on=['phonetic_first','phonetic_last','state','city'])

In [137]:
merge_phonetic_name_location_city['inventor_id'].nunique()

190

<span style="color:red">**Checkpoint 3: Deterministic matching of organization names**</span>

Now try doing different types of merges using organization names, blocking on different variables and seeing the differences.

In [149]:
assignees['org_city'] = assignees['org_city'].str.upper()
assignees.head()

patent_id             organization   org_city org_state org_country  \
0   9992413                 raytheon  LEXINGTON        MA          US   
1   RE45718        boston scientific  LEXINGTON        MA          US   
2   9992413                 raytheon    WALTHAM        MA          US   
3   9988932  honeywell international    WALTHAM        MA          US   
4   9991578  honeywell international    WALTHAM        MA          US   

           org_phonetic  
0                RAYTAN  
1       BASTANSCANTAFAC  
2                RAYTAN  
3  HANAYWALANTARNATANAL  
4  HANAYWALANTARNATANAL

In [158]:
organizations = organizations.rename(columns={' ORGANIZATION_STATE':'org_state'})
organizations = organizations.rename(columns={' ORGANIZATION_CITY':'org_city'})
organizations = organizations.rename(columns={' ORGANIZATION_NAME':'organization'})

In [159]:
organizations.head()

PROJECT_ID                         organization    org_city org_state  \
0      121219                university of arizona      TUCSON        AZ   
1      121220             ohio aerospace institute  BROOK PARK        OH   
2      121221              northwestern university    EVANSTON        IL   
3      121222  university of california santa cruz  SANTA CRUZ        CA   
4      121223      university of california irvine      IRVINE        CA   

   ORGANIZATION_COUNTRY                  org_phonetic  
0         UNITED STATES            UNAVARSATYAFARASAN  
1         UNITED STATES             OHARASPACANSTATAT  
2         UNITED STATES         NARTWASTARNANAVARSATY  
3         UNITED STATES  UNAVARSATYAFCALAFARNASANTACR  
4         UNITED STATES     UNAVARSATYAFCALAFARNARVAN

In [160]:
#Merging with organization names only
mrg = assignees.merge(organizations,on=['organization'])
print(mrg['patent_id'].nunique())
print(mrg['patent_id'].nunique()/assignees['patent_id'].nunique())

191
0.3281786941580756


In [153]:
#Merging with organization names only
merge_org = assignees.merge(organizations,on=['org_phonetic'])
print(merge_org['patent_id'].nunique())
print(merge_org['patent_id'].nunique()/assignees['patent_id'].nunique())

195
0.33505154639175255


In [155]:
#Merging with organization names with state
merge_org_st = assignees.merge(organizations,on=['org_phonetic','org_state'])
print(merge_org_st['patent_id'].nunique())
print(merge_org_st['patent_id'].nunique()/assignees['patent_id'].nunique())

193
0.3316151202749141


In [161]:
#Merging with organization names with state and city
merge_org_st_city = assignees.merge(organizations,on=['org_phonetic','org_state','org_city'])
print(merge_org_st_city['patent_id'].nunique())

#How many patents issued bwt 2015-2019 are from organizations that received grants in 2010?
print(merge_org_st_city['patent_id'].nunique()/assignees['patent_id'].nunique())

171
0.29381443298969073


In [163]:
print(assignees['org_phonetic'].nunique())
print(organizations['org_phonetic'].nunique())
print(merge_org_st_city['org_phonetic'].nunique())

280
5486
75


**Optional: Regular Expressions - regex**

When defining a regular expression search pattern, it is a good idea to start out by writing down, explicitly, in plain English, what you are trying to search for and exactly how you identify when you've found a match.
For example, if we look at an author field formatted as "<last_name> , <first_name> <middle_name>", in plain English, this is how I would explain where to find the last name: "starting from the beginning of the line, take all the characters until you see a comma."

We can build a regular expression that captures this idea from the following components:
- ^ Matches beginning of the line
- . Matches any character
- .+ A modifier that means "match one or more of the preceding expression"

In a regular expression, there are special reserved characters and character classes like those in the list above. Anything that is not a special character or class is just looked for explicitly (for example, a comma is not a special character in regular expressions, so if it is in a regular expression pattern, the regular expression processor will just be looking for a comma in the string, at that point in the pattern).

Note: if you want to actually look for one of these reserved characters, it must be escaped, so that, for example, the expression looks for a literal period, rather than the special regular expression meaning of a period. To escape a reserved character in a regular expression, precede it with a back slash ( "." ).
This results in the regular expression: ^.+,

We start at the beginning of the line ( "^" ), matching any characters ( ".+" ) until we come to the literal character of a comma ( "," ).

In python, to use a regular expression like this to search for matches in a given string, we use the built-in "re" package ( https://docs.python.org/2/library/re.html ), specifically the "re.search()" method. To use "re.search()", pass it first the regular expression you want to use to search, enclosed in quotation marks, and then the string you want to search within. 

#### REGEX CHEATSHEET


    - abc...     Letters
    - 123...     Digits
    - \d         Any Digit
    - \D         Any non-Digit Character
    - .          Any Character
    - \.         Period
    - [a,b,c]    Only a, b or c
    - [^a,b,c]   Not a,b, or c
    - [a-z]      Characters a to z
    - [0-9]      Numbers 0 to 9
    - \w any     Alphanumeric chracter
    - \W         any non-Alphanumeric character
    - {m}        m Repetitions
    - {m,n}      m to n repetitions
    - *          Zero or more repetitions
    - +          One or more repetitions
    - ?          Optional Character
    - \s         any Whitespace
    - \S         any non-Whitespace character
    - ^...$      Starts & Ends
    - (...)      Capture Group
    - (a(bc))    Capture sub-Group
    - (.*)       Capture All
    - (abc|def)  Capture abc or def
     
#### EXAMPLES
    - (\d\d|\D) will match 22X, 23G, 56H, etc...
    - \w will match any characters between 0-9 or a-z
    - \w{1-3} will match any alphanumeric character of a length of 1 to 3. 
    - (spell|spells) will match spell or spells

## References and Further Readings


### Record Linkage
Winkler, William E. 2009. “Record Linkage.” In Handbook of Statistics 29a, Sample Surveys: Design, Methods and Applications, edited by D. Pfeffermann and C.
R. Rao, 351–80. Elsevier.

Lane, Julia, Ian Foster, Rayid Ghani, Ron S. Jarmin, Frauke Kreuter (editors), Big Data and Social Science: A Practical Guide to Methods and Tools, Chapman and Hall/CRC Press, 2016. 
https://coleridge-initiative.github.io/big-data-and-social-science/chap-link.html

### Parsing

* Python online documentation: https://docs.python.org/2/library/string.html#deprecated-string-functions
* Python 2.7 Tutorial (Splitting and Joining Strings): http://www.pitt.edu/~naraehan/python2/split_join.html

### Regular Expression

* Python documentation: https://docs.python.org/2/library/re.html#regular-expression-syntax
* Online regular expression tester (good for learning): http://regex101.com/

### Record Linkage Python package
* `recordlinkage` Python package: https://recordlinkage.readthedocs.io/en/latest/index.html